src/ngraph/ngraph_graph.cc

```diff
@@ -262,96 +269,148 @@ std::vector<NodePtr> FindSubgraph(const Graph& graph, NodePtr node,
   int sg = 1;
-  // loop over the nodes from the back
-  for (auto n = graph.nodes_.rbegin(); n != graph.nodes_.rend(); ++n) {
-    if ((*n)->subgraph_ == 0) {
-      // select nodes in the a subgraph starting here and going up the graph
-      auto subgraph_nodes = FindSubgraph(graph, *n, func);
-
-      // if we found a significantly large subgraph, label it
-      if (subgraph_nodes.size() > 0) {
-        for (auto node : subgraph_nodes) {
-          node->subgraph_ = sg;
-        }
-        for (auto node : subgraph_nodes)
-          for (auto i : node->inputs_) {
-            if (i->subgraph_ != sg) i->subgraph_ = -1;
-          }
 
+  // collapse graphs from the outputs
+  for (size_t i = 0; i < graph->num_outputs_; ++i) {
+    bool found_subgraph = false;
+    found_subgraph = IdentifyOneSubgraph(graph, func, sg, graph->outputs_[i]);
+    if (found_subgraph) {
+      sg += 1;
+    }
+  }
+
+  // loop over the nodes from the back and find any that aren't in subgraphs
+  while (true) {
+    bool found_subgraph = false;
+    for (auto n = graph->nodes_.rbegin(); n != graph->nodes_.rend(); ++n) {
+      found_subgraph = IdentifyOneSubgraph(graph, func, sg, *n);
+      if (found_subgraph) {
         sg += 1;
+        break;
       }
     }
+    if (found_subgraph) {
+      continue;
+    } else {
+      break;
+    }
   }
 }
 
 // Function to collapse the intermediary graph into a graph
 // with subgraphs for nodes
-void CollapseSubgraphs(Graph* graph) {
+void CollapseSubgraph(Graph* graph, int subgraph_num) {
   // loop variable for undefined number of subgraphs
-  int i = 1;
-  while (true) {
-    auto tmpGraph = std::make_shared<Graph>(
-        "subgraph_" + randomString(12) + std::to_string(i), graph->context_);
-    // loop over all nodes and add nodes in the current subgraph to
-    for (auto node : graph->nodes_)
-      if (node->subgraph_ == i) tmpGraph->AddNode(node);
-
-    if (tmpGraph->nodes_.size() == 0) {
-      // if we don't find any nodes, assume we've run out of subgraphs
-      break;
-    } else {
-      // if we found nodes, setup subgraph
-      tmpGraph->in_ngraph_ = true;
-      tmpGraph->subgraph_ = i;
-      // set node name and shape based on last node in the subgraph
-      auto output = tmpGraph->nodes_.back();
-      tmpGraph->shape_ = output->shape_;
-      tmpGraph->dtype_ = output->dtype_;
-      tmpGraph->stype_ = output->stype_;
-
-      auto in_tmpGraphInputs = [&tmpGraph](NodePtr n) {
-        if (!in_vec(tmpGraph->inputs_, n)) return false;
-        return true;
-      };
-      // setup inputs to this subgraph (as a node)
-      for (auto node : tmpGraph->nodes_) {
-        for (auto input : node->inputs_) {
-          if (input->subgraph_ != i && !in_tmpGraphInputs(input))
-            tmpGraph->inputs_.emplace_back(input);
+  auto tmpGraph = std::make_shared<Graph>(
+      "subgraph_" + randomString(12) + std::to_string(subgraph_num),
```

Comment:
Why is random part of this algorithm?  At first blush it seems really evil.

Reply:
Because we have to register the graph with NNVM and it has to have a unique name.

Reply:
But why not a serial number or something?  Using random here seems strictly more confusing to me, plus there's a small chance of actually getting a duplicate value.

Reply:
Very small, like 1 of every 7e38 random strings will be the same.

We could theoretically keep a static increment around or something such, but high entropy randomness has always seemed the most appropriate to me. Probably too many Monte Carlo simulations in my past.